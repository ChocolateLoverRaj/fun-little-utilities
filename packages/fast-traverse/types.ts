export type MaybePromise<Value> = Value | Promise<Value>

export interface StatReturn {
  readonly isDirectory: () => boolean
}

export interface DeepParam<Path, DirName, BaseName> {
  level: number
  basename: BaseName
  path: Path
  container: DirName
}

export interface Options<
  DirectoryList extends Iterable<BaseName>,
  Path = string,
  BaseName = Path
> {
  readonly dirname: Path
  readonly readdir: (dirname: Path) => MaybePromise<DirectoryList>
  readonly stat: (path: Path) => MaybePromise<StatReturn>
  readonly join: (container: Path, basename: BaseName) => Path
  readonly deep?: (param: DeepParam<Path, Path, BaseName>) => MaybePromise<boolean>
  readonly level?: number
}
