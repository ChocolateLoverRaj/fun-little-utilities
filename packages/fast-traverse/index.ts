export * from './types'
export * from './traverse'
export { default } from './traverse'
