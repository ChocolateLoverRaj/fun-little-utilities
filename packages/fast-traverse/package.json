{
  "name": "fast-traverse",
  "version": "0.1.6",
  "description": "Traverse a FileSystem tree as fast as possible",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "browser": "index.mjs",
  "module": "index.mjs",
  "main": "index.js",
  "dependencies": {
    "tslib": "^1.11.1",
    "@types/node": "^13.11.0"
  }
}
