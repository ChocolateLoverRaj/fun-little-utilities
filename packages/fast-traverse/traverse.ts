import { Options } from './types'

export const DEFAULT_DEEP = () => true as const

export async function * traverse<
  DirectoryList extends Iterable<BaseName>,
  Path = string,
  BaseName = Path
> (options: Options<DirectoryList, Path, BaseName>): AsyncGenerator<DirectoryList> {
  const {
    dirname,
    readdir,
    stat,
    join,
    deep = DEFAULT_DEEP,
    level = 0
  } = options

  const dirs = await readdir(dirname)
  yield dirs

  for (const basename of dirs) {
    const path = join(dirname, basename)
    const statReturn = await stat(path)
    if (!statReturn.isDirectory()) continue

    const deepReturn = await deep({
      basename,
      path,
      level,
      container: dirname
    })
    if (!deepReturn) continue

    yield * traverse({
      readdir,
      stat,
      join,
      deep,
      dirname: path,
      level: level + 1
    })
  }
}

export default traverse
