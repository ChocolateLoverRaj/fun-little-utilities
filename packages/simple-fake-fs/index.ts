export enum ErrorKind {
  ENOENT = -2,
  EEXIST = -17,
  ENOTDIR = -20,
  EISDIR = -21,
}

export type ErrorCode = string & keyof typeof ErrorKind

export type SysCall =
  | 'stat'
  | 'scandir'
  | 'open'
  | 'read'
  | 'mkdir'
  | 'exists'

interface ErrorConstructor<Return, Path> {
  new (syscall: SysCall, path: Path): Return
}

export class ENOENT<Path> {
  public readonly errno = ErrorKind.ENOENT
  public readonly code = ErrorKind[ErrorKind.ENOENT]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export class EEXIST<Path> {
  public readonly errno = ErrorKind.EEXIST
  public readonly code = ErrorKind[ErrorKind.EEXIST]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export class ENOTDIR<Path> {
  public readonly errno = ErrorKind.ENOTDIR
  public readonly code = ErrorKind[ErrorKind.ENOTDIR]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export class EISDIR<Path> {
  public readonly errno = ErrorKind.EISDIR
  public readonly code = ErrorKind[ErrorKind.EISDIR]
  constructor(
    public readonly syscall: SysCall,
    public readonly path: Path,
  ) {}
}

export enum ContentKind {
  File = 'File',
  Directory = 'Directory',
  None = 'None',
  Error = 'Error',
}

const EMPTY_CLASS = class {}

class None {
  public readonly kind = ContentKind.None
}

const NONE = new None()

class ErrorCarrier<Value> {
  public readonly kind = ContentKind.Error
  constructor(public readonly value: Value) {}
}

export class FakeStats<FS extends Content<any, any>> {
  constructor(private readonly fs: FS) {}
  public readonly isFile = () => this.fs.kind === ContentKind.File
  public readonly isDirectory = () => this.fs.kind === ContentKind.Directory
}

export class FakeFileContent<Content> {
  public readonly kind = ContentKind.File
  constructor(public readonly content: Content) {}
}

interface FileSystemEntry<PathElm, FileContent> extends ReadonlyArray<any> {
  readonly 0: PathElm
  readonly 1: Content<PathElm, FileContent> | readonly FileSystemEntry<PathElm, FileContent>[]
}

export class FakeDirectoryContent<PathElm, FileContent> extends Map<
  PathElm,
  Content<PathElm, FileContent>,
> {
  public readonly kind = ContentKind.Directory

  constructor(entries: readonly FileSystemEntry<PathElm, FileContent>[] = []) {
    super(entries.map(([key, value]) => {
      if (value instanceof FakeFileContent || value instanceof FakeDirectoryContent) {
        return [key, value]
      }

      const newValue = new FakeDirectoryContent(value)
      return [key, newValue] as const
    }))
  }

  public hasPath(path: readonly PathElm[]) {
    const { kind } = this.getPath(path, 'exists', EMPTY_CLASS, EMPTY_CLASS)
    return kind === ContentKind.File || kind === ContentKind.Directory
  }

  public getPath<ERP, EFD>(
    path: readonly PathElm[],
    syscall: SysCall,
    RedundantPath: ErrorConstructor<ERP, PathElm[]>,
    FileAsDir: ErrorConstructor<EFD, PathElm[]>,
    original = () => Array.from(path),
  ): MaybeContent<
    PathElm,
    FileContent,
    ERP | EFD
  > {
    if (!path.length) return this
    const [first, ...rest] = path
    const next = this.get(first)
    if (!next) {
      return rest.length ? new ErrorCarrier(new RedundantPath(syscall, original())) : NONE
    }
    if (next.kind !== ContentKind.Directory) {
      return rest.length ? new ErrorCarrier(new FileAsDir(syscall, original())) : next
    }
    return next.getPath(rest, syscall, RedundantPath, FileAsDir, original)
  }

  public setPath<ENE, EFD>(
    path: readonly PathElm[],
    value: Content<PathElm, FileContent>,
    syscall: SysCall,
    NotExist: ErrorConstructor<ENE, PathElm[]>,
    FileAsDir: ErrorConstructor<EFD, PathElm[]>,
    original = () => Array.from(path),
  ): ENE | EFD | null {
    const [first, ...rest] = path
    if (rest.length) {
      const child = this.get(first)
      if (!child) return new NotExist(syscall, original())
      if (child.kind !== ContentKind.Directory) return new FileAsDir(syscall, original())
      return child.setPath(rest, value, syscall, NotExist, FileAsDir, original)
    }
    this.set(first, value)
    return null
  }

  public ensurePath<EFD>(
    path: readonly PathElm[],
    value: Content<PathElm, FileContent>,
    syscall: SysCall,
    FileAsDir: ErrorConstructor<EFD, PathElm[]>,
    original = () => Array.from(path),
  ): EFD | null {
    const [first, ...rest] = path
    if (rest.length) {
      const child = this.get(first)
      if (!child) {
        const next = new FakeDirectoryContent<PathElm, FileContent>()
        const error = next.ensurePath(rest, value, syscall, FileAsDir, original)
        if (error) return error
        this.set(first, next)
        return null
      }
      if (child.kind !== ContentKind.Directory) {
        return new FileAsDir(syscall, original())
      }
      return child.ensurePath(rest, value, syscall, FileAsDir, original)
    }
    this.set(first, value)
    return null
  }
}

export type Content<PathElm, FileContent> =
  | FakeFileContent<FileContent>
  | FakeDirectoryContent<PathElm, FileContent>

type MaybeContent<PathElm, FileContent, ErrorValue> =
  | None
  | ErrorCarrier<ErrorValue>
  | Content<PathElm, FileContent>

export class ArrayPathFileSystem<PathElm, FileContent> {
  private coreMap: FakeDirectoryContent<PathElm, FileContent>

  constructor(entries?: readonly FileSystemEntry<PathElm, FileContent>[]) {
    this.coreMap = new FakeDirectoryContent(entries)
  }

  public existsSync = (path: readonly PathElm[]) => this.coreMap.hasPath(path)

  public statSync = (path: readonly PathElm[]) => {
    const content = this.coreMap.getPath(path, 'stat', ENOENT, ENOTDIR)
    switch (content.kind) {
      case ContentKind.None:
        throw new ENOENT('stat', path)
      case ContentKind.Error:
        throw content.value
    }
    return new FakeStats(content)
  }

  public readdirSync = (dirname: readonly PathElm[]) => {
    const content = this.coreMap.getPath(dirname, 'scandir', ENOENT, ENOTDIR)
    switch (content.kind) {
      case ContentKind.Directory:
        return Array.from(content.keys())
      case ContentKind.File:
        throw new ENOTDIR('scandir', dirname)
      case ContentKind.None:
        throw new ENOENT('scandir', dirname)
      case ContentKind.Error:
        throw content.value
    }
  }

  public readFileSync = (filename: readonly PathElm[]) => {
    const content = this.coreMap.getPath(filename, 'read', ENOENT, ENOTDIR)
    switch (content.kind) {
      case ContentKind.File:
        return content.content
      case ContentKind.Directory:
        throw new EISDIR('read', filename)
      case ContentKind.None:
        throw new ENOENT('read', filename)
      case ContentKind.Error:
        throw content.value
    }
  }

  public mkdirSync = (dirname: readonly PathElm[]) => {
    switch (this.coreMap.getPath(dirname, 'mkdir', EMPTY_CLASS, EMPTY_CLASS).kind) {
      case ContentKind.File:
      case ContentKind.Directory:
        throw new EEXIST('mkdir', dirname)
    }
    const error = this.coreMap.setPath(dirname, new FakeDirectoryContent(), 'mkdir', ENOENT, ENOTDIR)
    if (error) throw error
  }

  public writeFileSync = (filename: readonly PathElm[], fileContent: FileContent) => {
    if (this.coreMap.getPath(filename, 'open', EMPTY_CLASS, EMPTY_CLASS).kind === ContentKind.Directory) {
      throw new EISDIR('open', filename)
    }
    const error = this.coreMap.setPath(filename, new FakeFileContent(fileContent), 'open', ENOENT, ENOTDIR)
    if (error) throw error
  }

  public ensureDirSync = (dirname: readonly PathElm[]) => {
    switch (this.coreMap.getPath(dirname, 'mkdir', ENOENT, ENOTDIR).kind) {
      case ContentKind.File:
        throw new ENOTDIR('mkdir', dirname)
      case ContentKind.Directory:
        return
    }
    const error = this.coreMap.ensurePath(dirname, new FakeDirectoryContent(), 'mkdir', ENOTDIR)
    if (error) throw error
  }

  public outputFileSync = (filename: readonly PathElm[], content: FileContent) => {
    if (this.coreMap.getPath(filename, 'open', ENOENT, ENOTDIR).kind === ContentKind.Directory) {
      throw new EISDIR('open', filename)
    }
    const error = this.coreMap.ensurePath(filename, new FakeFileContent(content), 'open', ENOTDIR)
    if (error) throw error
  }
}

type StringPathFileSystemDict = {
  readonly [_: string]: StringPathFileSystemDict | string
}

function dict2entries(dict: StringPathFileSystemDict): FileSystemEntry<string, string>[] {
  return Object.entries(dict).map(([key, value]) => {
    if (typeof value === 'string') {
      return [key, new FakeFileContent(value)]
    }

    return [key, dict2entries(value)]
  })
}

export class StringPathFileSystem {
  private readonly core: ArrayPathFileSystem<string, string>

  constructor(
    private readonly sep: string,
    dict: StringPathFileSystemDict = {},
  ) {
    this.core = new ArrayPathFileSystem(dict2entries(dict))
  }

  private normalize(item: string) {
    return item === '.' ? '' : item
  }

  private split(path: string) {
    return path
      .split(this.sep)
      .map(this.normalize)
      .filter(Boolean)
  }

  public existsSync = (path: string) => this.core.existsSync(this.split(path))

  public statSync = (path: string) => this.core.statSync(this.split(path))

  public readdirSync = (path: string) => this.core.readdirSync(this.split(path))

  public mkdirSync = (path: string) => this.core.mkdirSync(this.split(path))

  public readFileSync = (path: string) => this.core.readFileSync(this.split(path))

  public writeFileSync = (path: string, fileContent: string) => this.core.writeFileSync(this.split(path), fileContent)

  public ensureDirSync = (path: string) => this.core.ensureDirSync(this.split(path))

  public outputFileSync = (path: string, fileContent: string) => this.core.outputFileSync(this.split(path), fileContent)

  public ensureFileSync = (path: string) => this.outputFileSync(path, '')
}
