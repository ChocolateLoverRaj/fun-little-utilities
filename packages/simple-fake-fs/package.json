{
  "name": "simple-fake-fs",
  "version": "0.0.7",
  "description": "Fake FS-like tree",
  "author": "Hoàng Văn Khải <hvksmr1996@gmail.com>",
  "license": "MIT",
  "homepage": "https://ksxnodeapps.github.io/fun-little-utilities/",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/ksxnodeapps/fun-little-utilities.git"
  },
  "bugs": {
    "url": "https://github.com/ksxnodeapps/fun-little-utilities/issues"
  },
  "dependencies": {
    "tslib": "^1.10.0",
    "@types/node": "^13.5.2"
  }
}
