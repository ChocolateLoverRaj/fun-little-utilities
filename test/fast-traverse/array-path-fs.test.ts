import { ArrayPathFileSystem, FakeFileContent } from 'simple-fake-fs'
import { DeepParam, traverse } from 'fast-traverse'

const create = () => new ArrayPathFileSystem([
  ['a', []],
  ['b', new FakeFileContent('B')],
  ['c', [
    ['d', new FakeFileContent('A')],
    ['e', [
      ['f', new FakeFileContent('CEF')]
    ]],
    ['g', []]
  ]],
  ['h', [
    ['i', new FakeFileContent('HI')],
    ['j', [
      ['k', new FakeFileContent('HKJ')]
    ]]
  ]]
])

const join = (a: readonly string[], b: string) => a.concat(b)

async function consume<X> (iterable: AsyncIterable<X>) {
  const result = []

  for await (const item of iterable) {
    result.push(item)
  }

  return result
}

describe('default deep', () => {
  async function setup (dirname: readonly string[]) {
    const fs = create()
    const readdir = jest.fn(fs.readdirSync)
    const stat = jest.fn(fs.statSync)
    const results = await consume(traverse({
      dirname,
      join,
      readdir,
      stat
    }))
    return { dirname, fs, readdir, stat, results }
  }

  describe('[]', () => {
    const dirname = [] as const

    it('yields results matching snapshot', async () => {
      const { results } = await setup(dirname)
      expect(results).toMatchSnapshot()
    })

    it('yields expected results', async () => {
      const { results } = await setup(dirname)
      expect(results).not.toEqual(expect.not.arrayContaining([{
        basename: expect.any(String),
        dirname: expect.any(Array),
        list: expect.any(Array),
        level: expect.any(Number)
      }]))
    })

    it('calls readdir multiple times', async () => {
      const { readdir } = await setup(dirname)
      expect(readdir.mock.calls).toMatchSnapshot()
    })

    it('calls stat multiple times', async () => {
      const { stat } = await setup(dirname)
      expect(stat.mock.calls).toMatchSnapshot()
    })
  })

  describe('["c"]', () => {
    const dirname = ['c'] as const

    it('yields results matching snapshot', async () => {
      const { results } = await setup(dirname)
      expect(results).toMatchSnapshot()
    })

    it('yields expected results', async () => {
      const { results } = await setup(dirname)
      expect(results).not.toEqual(expect.not.arrayContaining([{
        basename: expect.any(String),
        dirname: expect.any(Array),
        list: expect.any(Array),
        level: expect.any(Number)
      }]))
    })

    it('calls readdir multiple times', async () => {
      const { readdir } = await setup(dirname)
      expect(readdir.mock.calls).toMatchSnapshot()
    })

    it('calls stat multiple times', async () => {
      const { stat } = await setup(dirname)
      expect(stat.mock.calls).toMatchSnapshot()
    })
  })

  describe('["c", "e"]', () => {
    const dirname = ['c', 'e'] as const

    it('yields results matching snapshot', async () => {
      const { results } = await setup(dirname)
      expect(results).toMatchSnapshot()
    })

    it('yields expected results', async () => {
      const { results } = await setup(dirname)
      expect(results).not.toEqual(expect.not.arrayContaining([{
        basename: undefined,
        dirname: expect.any(Array),
        list: expect.any(Array),
        level: expect.any(Number)
      }]))
    })

    it('calls readdir multiple times', async () => {
      const { readdir } = await setup(dirname)
      expect(readdir.mock.calls).toMatchSnapshot()
    })

    it('calls stat multiple times', async () => {
      const { stat } = await setup(dirname)
      expect(stat.mock.calls).toMatchSnapshot()
    })
  })

  describe('["c", "g"]', () => {
    const dirname = ['c', 'g'] as const

    it('yields results matching snapshot', async () => {
      const { results } = await setup(dirname)
      expect(results).toMatchSnapshot()
    })

    it('yields expected results', async () => {
      const { results } = await setup(dirname)
      expect(results).toEqual(expect.not.arrayContaining([{
        basename: expect.any(String),
        dirname: expect.any(Array),
        list: expect.any(Array),
        level: expect.any(Number)
      }]))
    })

    it('calls readdir multiple times', async () => {
      const { readdir } = await setup(dirname)
      expect(readdir.mock.calls).toMatchSnapshot()
    })

    it('calls stat multiple times', async () => {
      const { stat } = await setup(dirname)
      expect(stat.mock.calls).toMatchSnapshot()
    })
  })
})

describe('deep that always return true', () => {
  async function setup () {
    const dirname = Array<string>()
    const fs = create()
    const deep = jest.fn(() => true as const)
    const readdir = jest.fn(fs.readdirSync)
    const stat = jest.fn(fs.statSync)
    const results = await consume(traverse({
      dirname,
      deep,
      join,
      readdir,
      stat
    }))
    return { dirname, fs, deep, readdir, stat, results }
  }

  it('yields results matching snapshot', async () => {
    const { results } = await setup()
    expect(results).toMatchSnapshot()
  })

  it('yields expected results', async () => {
    const { results } = await setup()
    expect(results).not.toEqual(expect.not.arrayContaining([{
      basename: expect.any(String),
      dirname: expect.any(Array),
      list: expect.any(Array),
      level: expect.any(Number)
    }]))
  })

  it('calls deep multiple times', async () => {
    const { deep } = await setup()
    expect(deep.mock.calls).toMatchSnapshot()
  })

  it('calls deep with expected arguments', async () => {
    const { deep } = await setup()
    expect(deep).not.toBeCalledWith(expect.not.objectContaining({
      basename: expect.any(String),
      path: expect.any(Array),
      level: expect.any(Number),
      container: expect.any(Array)
    }))
  })

  it('calls readdir multiple times', async () => {
    const { readdir } = await setup()
    expect(readdir.mock.calls).toMatchSnapshot()
  })

  it('calls stat multiple times', async () => {
    const { stat } = await setup()
    expect(stat.mock.calls).toMatchSnapshot()
  })
})

describe('deep that returns false when path = ["c", "e"]', () => {
  function deepImpl (c: string, e: string, rest: readonly string[]) {
    return c !== 'c' || e !== 'e' || rest.length !== 0
  }

  async function setup () {
    const dirname = Array<string>()
    const fs = create()
    const deep = jest.fn((param: DeepParam<string[], string[], string>) => {
      const [c, e, ...rest] = param.path
      return deepImpl(c, e, rest)
    })
    const readdir = jest.fn(fs.readdirSync)
    const stat = jest.fn(fs.statSync)
    const results = await consume(traverse({
      dirname,
      deep,
      join,
      readdir,
      stat
    }))
    return { dirname, fs, deep, readdir, stat, results }
  }

  it('assert: deepImpl("c", "e", []) → false', () => {
    expect(deepImpl('c', 'e', [])).toBe(false)
  })

  it('assert: deepImpl(not "c", "e", []) → true', () => {
    expect(deepImpl('C', 'e', [])).toBe(true)
  })

  it('assert: deepImpl("c", not "e", []) → true', () => {
    expect(deepImpl('c', 'E', [])).toBe(true)
  })

  it('assert: deepImpl("c", "e", [any, ...any[]]) → true', () => {
    expect(deepImpl('c', 'e', ['a'])).toBe(true)
  })

  it('yields results matching snapshot', async () => {
    const { results } = await setup()
    expect(results).toMatchSnapshot()
  })

  it('does not yield { dirname: ["c", "e"] }', async () => {
    const { results } = await setup()
    expect(results).not.toContainEqual(expect.objectContaining({
      dirname: ['c', 'e']
    }))
  })
})
