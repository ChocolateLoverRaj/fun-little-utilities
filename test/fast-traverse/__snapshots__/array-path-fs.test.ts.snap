// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep that always return true calls deep multiple times 1`] = `
Array [
  Array [
    Object {
      "basename": "a",
      "container": Array [],
      "level": 0,
      "path": Array [
        "a",
      ],
    },
  ],
  Array [
    Object {
      "basename": "c",
      "container": Array [],
      "level": 0,
      "path": Array [
        "c",
      ],
    },
  ],
  Array [
    Object {
      "basename": "e",
      "container": Array [
        "c",
      ],
      "level": 1,
      "path": Array [
        "c",
        "e",
      ],
    },
  ],
  Array [
    Object {
      "basename": "g",
      "container": Array [
        "c",
      ],
      "level": 1,
      "path": Array [
        "c",
        "g",
      ],
    },
  ],
  Array [
    Object {
      "basename": "h",
      "container": Array [],
      "level": 0,
      "path": Array [
        "h",
      ],
    },
  ],
  Array [
    Object {
      "basename": "j",
      "container": Array [
        "h",
      ],
      "level": 1,
      "path": Array [
        "h",
        "j",
      ],
    },
  ],
]
`;

exports[`deep that always return true calls readdir multiple times 1`] = `
Array [
  Array [
    Array [],
  ],
  Array [
    Array [
      "a",
    ],
  ],
  Array [
    Array [
      "c",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
    ],
  ],
  Array [
    Array [
      "c",
      "g",
    ],
  ],
  Array [
    Array [
      "h",
    ],
  ],
  Array [
    Array [
      "h",
      "j",
    ],
  ],
]
`;

exports[`deep that always return true calls stat multiple times 1`] = `
Array [
  Array [
    Array [
      "a",
    ],
  ],
  Array [
    Array [
      "b",
    ],
  ],
  Array [
    Array [
      "c",
    ],
  ],
  Array [
    Array [
      "c",
      "d",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
      "f",
    ],
  ],
  Array [
    Array [
      "c",
      "g",
    ],
  ],
  Array [
    Array [
      "h",
    ],
  ],
  Array [
    Array [
      "h",
      "i",
    ],
  ],
  Array [
    Array [
      "h",
      "j",
    ],
  ],
  Array [
    Array [
      "h",
      "j",
      "k",
    ],
  ],
]
`;

exports[`deep that always return true yields results matching snapshot 1`] = `
Array [
  Object {
    "basename": undefined,
    "dirname": Array [],
    "level": 0,
    "list": Array [
      "a",
      "b",
      "c",
      "h",
    ],
  },
  Object {
    "basename": "a",
    "dirname": Array [
      "a",
    ],
    "level": 1,
    "list": Array [],
  },
  Object {
    "basename": "c",
    "dirname": Array [
      "c",
    ],
    "level": 1,
    "list": Array [
      "d",
      "e",
      "g",
    ],
  },
  Object {
    "basename": "e",
    "dirname": Array [
      "c",
      "e",
    ],
    "level": 2,
    "list": Array [
      "f",
    ],
  },
  Object {
    "basename": "g",
    "dirname": Array [
      "c",
      "g",
    ],
    "level": 2,
    "list": Array [],
  },
  Object {
    "basename": "h",
    "dirname": Array [
      "h",
    ],
    "level": 1,
    "list": Array [
      "i",
      "j",
    ],
  },
  Object {
    "basename": "j",
    "dirname": Array [
      "h",
      "j",
    ],
    "level": 2,
    "list": Array [
      "k",
    ],
  },
]
`;

exports[`deep that returns false when path = ["c", "e"] yields results matching snapshot 1`] = `
Array [
  Object {
    "basename": undefined,
    "dirname": Array [],
    "level": 0,
    "list": Array [
      "a",
      "b",
      "c",
      "h",
    ],
  },
  Object {
    "basename": "a",
    "dirname": Array [
      "a",
    ],
    "level": 1,
    "list": Array [],
  },
  Object {
    "basename": "c",
    "dirname": Array [
      "c",
    ],
    "level": 1,
    "list": Array [
      "d",
      "e",
      "g",
    ],
  },
  Object {
    "basename": "g",
    "dirname": Array [
      "c",
      "g",
    ],
    "level": 2,
    "list": Array [],
  },
  Object {
    "basename": "h",
    "dirname": Array [
      "h",
    ],
    "level": 1,
    "list": Array [
      "i",
      "j",
    ],
  },
  Object {
    "basename": "j",
    "dirname": Array [
      "h",
      "j",
    ],
    "level": 2,
    "list": Array [
      "k",
    ],
  },
]
`;

exports[`default deep ["c", "e"] calls readdir multiple times 1`] = `
Array [
  Array [
    Array [
      "c",
      "e",
    ],
  ],
]
`;

exports[`default deep ["c", "e"] calls stat multiple times 1`] = `
Array [
  Array [
    Array [
      "c",
      "e",
      "f",
    ],
  ],
]
`;

exports[`default deep ["c", "e"] yields results matching snapshot 1`] = `
Array [
  Object {
    "basename": undefined,
    "dirname": Array [
      "c",
      "e",
    ],
    "level": 0,
    "list": Array [
      "f",
    ],
  },
]
`;

exports[`default deep ["c", "g"] calls readdir multiple times 1`] = `
Array [
  Array [
    Array [
      "c",
      "g",
    ],
  ],
]
`;

exports[`default deep ["c", "g"] calls stat multiple times 1`] = `Array []`;

exports[`default deep ["c", "g"] yields results matching snapshot 1`] = `
Array [
  Object {
    "basename": undefined,
    "dirname": Array [
      "c",
      "g",
    ],
    "level": 0,
    "list": Array [],
  },
]
`;

exports[`default deep ["c"] calls readdir multiple times 1`] = `
Array [
  Array [
    Array [
      "c",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
    ],
  ],
  Array [
    Array [
      "c",
      "g",
    ],
  ],
]
`;

exports[`default deep ["c"] calls stat multiple times 1`] = `
Array [
  Array [
    Array [
      "c",
      "d",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
      "f",
    ],
  ],
  Array [
    Array [
      "c",
      "g",
    ],
  ],
]
`;

exports[`default deep ["c"] yields results matching snapshot 1`] = `
Array [
  Object {
    "basename": undefined,
    "dirname": Array [
      "c",
    ],
    "level": 0,
    "list": Array [
      "d",
      "e",
      "g",
    ],
  },
  Object {
    "basename": "e",
    "dirname": Array [
      "c",
      "e",
    ],
    "level": 1,
    "list": Array [
      "f",
    ],
  },
  Object {
    "basename": "g",
    "dirname": Array [
      "c",
      "g",
    ],
    "level": 1,
    "list": Array [],
  },
]
`;

exports[`default deep [] calls readdir multiple times 1`] = `
Array [
  Array [
    Array [],
  ],
  Array [
    Array [
      "a",
    ],
  ],
  Array [
    Array [
      "c",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
    ],
  ],
  Array [
    Array [
      "c",
      "g",
    ],
  ],
  Array [
    Array [
      "h",
    ],
  ],
  Array [
    Array [
      "h",
      "j",
    ],
  ],
]
`;

exports[`default deep [] calls stat multiple times 1`] = `
Array [
  Array [
    Array [
      "a",
    ],
  ],
  Array [
    Array [
      "b",
    ],
  ],
  Array [
    Array [
      "c",
    ],
  ],
  Array [
    Array [
      "c",
      "d",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
    ],
  ],
  Array [
    Array [
      "c",
      "e",
      "f",
    ],
  ],
  Array [
    Array [
      "c",
      "g",
    ],
  ],
  Array [
    Array [
      "h",
    ],
  ],
  Array [
    Array [
      "h",
      "i",
    ],
  ],
  Array [
    Array [
      "h",
      "j",
    ],
  ],
  Array [
    Array [
      "h",
      "j",
      "k",
    ],
  ],
]
`;

exports[`default deep [] yields results matching snapshot 1`] = `
Array [
  Object {
    "basename": undefined,
    "dirname": Array [],
    "level": 0,
    "list": Array [
      "a",
      "b",
      "c",
      "h",
    ],
  },
  Object {
    "basename": "a",
    "dirname": Array [
      "a",
    ],
    "level": 1,
    "list": Array [],
  },
  Object {
    "basename": "c",
    "dirname": Array [
      "c",
    ],
    "level": 1,
    "list": Array [
      "d",
      "e",
      "g",
    ],
  },
  Object {
    "basename": "e",
    "dirname": Array [
      "c",
      "e",
    ],
    "level": 2,
    "list": Array [
      "f",
    ],
  },
  Object {
    "basename": "g",
    "dirname": Array [
      "c",
      "g",
    ],
    "level": 2,
    "list": Array [],
  },
  Object {
    "basename": "h",
    "dirname": Array [
      "h",
    ],
    "level": 1,
    "list": Array [
      "i",
      "j",
    ],
  },
  Object {
    "basename": "j",
    "dirname": Array [
      "h",
      "j",
    ],
    "level": 2,
    "list": Array [
      "k",
    ],
  },
]
`;
